stream/stream_smb2.c

```diff
@@ -0,0 +1,440 @@
+        *(p->domain + domain_length)='\0';
+        MP_DBG(stream, "[smb] domain: %s\n", p->domain);
+
+        const size_t user_length = strlen(*(tokens + 1));
+        p->user = malloc(user_length + 1);
+        strncpy(p->user, *(tokens + 1), user_length);
+        *(p->user + user_length)='\0';
+        MP_DBG(stream, "[smb] user: %s\n", p->user);
+
+        const size_t password_length = strlen(*(tokens + 2));
+        p->password = malloc(password_length + 2);
+        strncpy(p->password, *(tokens + 2), password_length);
+        MP_DBG(stream, "[smb] password: %s\n", p->password);
+        *(p->password + password_length)='\0';
+    }
+    free(author_str);
+    if ((p->server) != NULL && (p->share) != NULL && (p->filepath) != NULL) {
+        return 0;
+    }
+    return -1;
+}
+
+static void free_priv(struct priv *p) {
+    if (p->domain) {
+        free(p->domain);
+    }
+
+    if (p->user) {
+        free(p->user);
+    }
+
+    if (p->password) {
+        free(p->password);
+    }
+
+    if (p->server) {
+        free(p->server);
+    }
+
+    if (p->share) {
+        free(p->share);
+    }
+
+    if (p->filepath) {
+        free(p->filepath);
+    }
+}
+
+static int control(stream_t *s, int cmd, void *arg) {
+    struct priv *p = s->priv;
+    switch(cmd) {
+        case STREAM_CTRL_GET_SIZE: {
+            pthread_mutex_lock(&smb2_lock);
+            *(int64_t *)arg = p->filesize;
+            pthread_mutex_unlock(&smb2_lock);
+            return 1;
+        }
+        break;
+    }
+    return STREAM_UNSUPPORTED;
+}
+
+static int seek(stream_t *s,int64_t newpos) {
+    struct priv *p = s->priv;
+    pthread_mutex_lock(&smb2_lock);
+    off_t current_offset;
+    off_t size = smb2_lseek(p->smb2_context, p->fh, newpos, SEEK_SET, &current_offset);
+    pthread_mutex_unlock(&smb2_lock);
+    if(size<0) {
+        return 0;
+    }
+    return 1;
+}
+
+static int fill_buffer(stream_t *s, char* buffer, int max_len){
+    struct priv *p = s->priv;
+    pthread_mutex_lock(&smb2_lock);
+    int r = smb2_read(p->smb2_context, p->fh, (void*)buffer, max_len);
+    pthread_mutex_unlock(&smb2_lock);
+    return r;
+    /*
+    int left_read = max_len;
+    int total_read = 0;
+    while (left_read > 0) {
+        char *pos = buffer + total_read;
+        int r = smb2_read(p->smb2_context, p->fh, (void*)pos, left_read);
+        if (r == 0) {
+            break;
+        }
+        total_read += r;
+        left_read -= r;
+    }
+    pthread_mutex_unlock(&smb2_lock);
+    return (total_read <= 0) ? -1 : total_read;
+     */
+}
+
+static int write_buffer(stream_t *s, char* buffer, int len) {
+    struct priv *p = s->priv;
+    int r = len;
+    int wr;
+    while (r > 0) {
+        pthread_mutex_lock(&smb2_lock);
+        wr = smb2_write(p->smb2_context, p->fh, (void*)buffer, r);
+        pthread_mutex_unlock(&smb2_lock);
+        if (wr <= 0)
+            return -1;
+        r -= wr;
+        buffer += wr;
+    }
+    return len - r;
+}
```

Comment:
I'll make stream.c handle short writes in master, so you can remove all this stuff (like in read).